client-v2/src/shared/util/createAsyncResourceBundle.js

```diff
@@ -0,0 +1,249 @@
+
+const createSelectors = (selectLocalState: (state: any) => any) => ({
+  selectCurrentError: createSelectCurrentError(selectLocalState),
+  selectLastError: createSelectLastError(selectLocalState),
+  selectLastFetch: createSelectLastFetch(selectLocalState),
+  selectIsLoading: createSelectIsLoading(selectLocalState),
+  selectIsLoadingById: createSelectIsLoadingById(selectLocalState),
+  selectById: createSelectById(selectLocalState),
+  selectAll: createSelectAll(selectLocalState)
+});
+
+type FetchStartAction = {|
+  // In this, and subsequent action types, we include a local
+  // type to ensure that Flow can disambiguate action types in the
+  // reducer. Alternative solutions welcome!
+  localType: 'START',
+  type: string,
+  payload: { ids?: string[] | string }
+|};
+
+type FetchSuccessAction = {|
+  localType: 'SUCCESS',
+  type: string,
+  payload: { data: any, time: number }
+|};
+
+type FetchErrorAction = {|
+  localType: 'ERROR',
+  type: string,
+  payload: {
+    error: string,
+    time: number,
+    ids?: string | string[]
+  }
+|};
+
+type Action = FetchStartAction | FetchSuccessAction | FetchErrorAction;
+
+type State<Resource> = {
+  data: Resource | { [id: string]: Resource },
+  lastError: string | null,
+  error: string | null,
+  lastFetch: number | null,
+  loadingIds: string[]
+};
+
+type BaseResource = {
+  id?: string
+};
+
+/**
+ * Creates a bundle of actions, selectors, and a reducer to handle
+ * common actions and selections for data that needs to be fetched:
+ * start, success and error actions, storing and selecting error states,
+ * and storing and selecting staleness data, as well as storing the
+ * fetched data itself.
+ *
+ * Consumers can add add their own actions and selectors, and extend
+ * the given reducer, to provide additional functionality.
+ */
+function createAsyncResourceBundle<Resource: BaseResource | any>(
+  // The name of the entity for which this reducer is responsible
+  entityName: string,
+  options: {|
+    // The key the reducer provided by this bundle is mounted at.
+    // Defaults to entityName if none is given.
+    selectLocalState?: (state: any) => any,
+    // Do we index the incoming data by id, or just add it to the state as-is?
+    indexById?: boolean,
+    // Provides a namespace for the created actions, separated by a slash,
+    // e.g.the resource 'books' namespaced with 'shared' becomes SHARED/BOOKS
+    namespace?: string,
+    // The initial state of the reducer data. Defaults to null.
+    initialData?: any
+  |} = {
+    indexById: false,
+    initialData: {}
+  }
+) {
+  const { indexById } = options;
+  const selectLocalState = options.selectLocalState
+    ? options.selectLocalState
+    : (state: any) => state[entityName];
+  const baseActionKey = snakeCase(entityName).toUpperCase();
+  const actionKey = options.namespace
+    ? `${options.namespace.toUpperCase()}/${baseActionKey}`
+    : baseActionKey;
+  const FETCH_START = `${actionKey}_FETCH_START`;
```

Comment:
Now I'm looking at this obviously you couldn't ask for this as a config as you need the string at compile time to discriminate the union. Will the explicit generic arguments help here when they come to babel?

Reply:
The factor re: @Ap0c 's suggestion should help here!

Reply:
Yeah sounds like the two of you sorted that riiight out 🎉